fabtests/common/hmem_synapseai.c

```diff
@@ -0,0 +1,446 @@
+		FT_ERR("Failed to synDeviceAcquireByModuleId()\n");
+		goto err;
+	}
+
+	if (synapseai_ops.synDeviceGetInfoV2(synapseai_fd, &deviceInfo) != synSuccess) {
+		FT_ERR("Failed to synDeviceGetInfoV2()\n");
+		goto err;
+	}
+	device_fd = deviceInfo.fd;
+
+	if (synapseai_ops.synStreamCreateGeneric(&synapseai_stream_handle, synapseai_fd, 0) !=
+		synSuccess) {
+		FT_ERR("Failed to synStreamCreateGeneric()\n");
+		goto err;
+	}
+
+	FT_DEBUG("Successfully initialized Synapseai");
+	return FI_SUCCESS;
+
+err:
+	cleanup_synapseai_ops();
+	return -FI_ENODATA;
+}
+
+int ft_synapseai_cleanup(void)
+{
+	if (synapseai_fd != -1) {
+		synapseai_ops.synDestroy();
+	}
+	cleanup_synapseai_ops();
+	return 0;
+}
+
+int ft_synapseai_alloc(uint64_t device, void **buf, size_t size)
+{
+	uint64_t addr;
+	size_t buf_size = (size + ACCEL_PAGE_SIZE - 1) & ~(ACCEL_PAGE_SIZE - 1);
+
+	if (synapseai_ops.synDeviceMalloc(synapseai_fd, buf_size, 0x0, 0, &addr) != synSuccess) {
+		FT_ERR("synDeviceMalloc failed");
+		return -FI_ENOBUFS;
+	}
+
+	if (addr == 0) {
+		FT_ERR("synDeviceMalloc returned invalid address");
+		return -FI_ENOBUFS;
+	}
+	*buf = (void *)addr;
+	return 0;
+}
+
+int ft_synapseai_free(void *buf)
+{
+	if (synapseai_ops.synDeviceFree(synapseai_fd, (uint64_t)buf, 0) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+	return 0;
+}
+
+int ft_synapseai_memset(uint64_t device, void *buf, int value, size_t size)
+{
+	if (synapseai_ops.synMemsetD8Async((uint64_t)buf, (unsigned char)value, size,
+										synapseai_stream_handle) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+
+	return stream_synchronize(synapseai_stream_handle);
+}
+
+int ft_synapseai_alloc_host(void **buf, size_t size)
+{
+	size_t buf_size = (size + ACCEL_PAGE_SIZE - 1) & ~(ACCEL_PAGE_SIZE - 1);
+
+	if (synapseai_ops.synHostMalloc(synapseai_fd, buf_size, 0, buf) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+	return FI_SUCCESS;
+}
+
+int ft_synapseai_free_host(void *buf)
+{
+	if (synapseai_ops.synHostFree(synapseai_fd, buf, 0) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+	return FI_SUCCESS;
+}
+
+int ft_synapseai_copy_to_hmem(uint64_t device, void *dst, const void *src, size_t size)
+{
+	if (synapseai_ops.synMemCopyAsync(synapseai_stream_handle, (uint64_t)src, size, (uint64_t)dst,
+									  HOST_TO_DRAM) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+	return stream_synchronize(synapseai_stream_handle);
+}
+
+int ft_synapseai_copy_from_hmem(uint64_t device, void *dst, const void *src, size_t size)
+{
+	if (synapseai_ops.synMemCopyAsync(synapseai_stream_handle, (uint64_t)src, size, (uint64_t)dst,
+									  DRAM_TO_HOST) != synSuccess) {
+		return -FI_ENOBUFS;
+	}
+	return stream_synchronize(synapseai_stream_handle);
+}
+
+int ft_synapseai_get_dmabuf_fd(void *buf, size_t len, int *dmabuf_fd, uint64_t *dmabuf_offset)
+{
+	scal_pool_handle_t mpHandle;
+	scal_memory_pool_infoV2 mpInfo;
+	scal_handle_t a = 0;
+
+	if (synapseai_ops.scal_get_handle_from_fd(device_fd, &a) != SCAL_SUCCESS) {
```

Comment:
Should we update https://github.com/ofiwg/libfabric/blob/main/src/hmem_synapseai.c#L192 ?

Reply:
We consider moving hcl code to use FI_MR_DMABUFF registration method. so currently it should not be updated